emcc.py

```diff
@@ -516,6 +517,58 @@ def get_all_js_syms():
     if shared.is_c_symbol(name):
       name = shared.demangle_c_symbol_name(name)
       library_syms.add(name)
+  return library_syms
+
+
+@ToolchainProfiler.profile_block('JS symbol generation')
+def get_all_js_syms():
+  # Avoiding using the cache when generating struct info since
+  # this step is performed while the cache is locked.
+  if settings.BOOTSTRAPPING_STRUCT_INFO or config.FROZEN_CACHE:
+    return generate_js_symbols()
+
+  # To avoid the cost of calling generate_js_symbols each time an executable is
+  # linked we cache symbol lists for the N most recently used configs.
+  # We define a cache hit as when the settings and `--js-library` contents are
+  # identical.
+  input_files = {}
+  input_files['settings.json'] = json.dumps(settings.dict(), sort_keys=True, indent=2)
+  for jslib in sorted(glob.glob(utils.path_from_root('src') + '/library*.js')):
+    input_files[jslib] = read_file(jslib)
+  for jslib in settings.JS_LIBRARIES:
+    if os.path.abspath(jslib):
+      jslib = utils.path_from_root('src', jslib)
+    input_files[jslib] = read_file(jslib)
+  input_data = []
```

Comment:
Might be worth asserting no js library has name `settings.json`, as that could lead to confusing behavior...

Reply:
I think that isn't possible, because all the other filenames are absolute .. I'll add a comment.